src/ptvsd/server/main.py

```diff
@@ -266,62 +266,68 @@ def run_code():
 
 
 def attach_to_pid():
-    def quoted_str(s):
-        if s is None:
-            return s
-        assert not isinstance(s, bytes)
-        unescaped = set(chr(ch) for ch in range(32, 127)) - {'"', "'", "\\"}
-
-        def escape(ch):
-            return ch if ch in unescaped else "\\u%04X" % ord(ch)
-
-        return 'u"' + "".join(map(escape, s)) + '"'
-
-    log.info("Attaching to process with ID {0}", options.target)
+    log.info('Attaching to process with ID {0}', options.target)
 
     pid = options.target
-    host = quoted_str(options.host)
+    host = options.host
     port = options.port
     client = options.client
-    multiprocess = options.multiprocess
-    log_dir = quoted_str(ptvsd.common.options.log_dir)
+    log_dir = common_opts.log_dir
+    if log_dir is None:
+        log_dir = ""
 
-    ptvsd_path = os.path.abspath(os.path.join(ptvsd.server.__file__, "../.."))
-    if isinstance(ptvsd_path, bytes):
-        ptvsd_path = ptvsd_path.decode(sys.getfilesystemencoding())
-    ptvsd_path = quoted_str(ptvsd_path)
+    attach_script_ptvsd_pid_dirname = os.path.join(os.path.dirname(ptvsd.__file__), 'common')
+    assert os.path.exists(attach_script_ptvsd_pid_dirname)
 
-    # pydevd requires injected code to not contain any single quotes.
-    code = """
-import os
-assert os.getpid() == {pid}
 
-import sys
-sys.path.insert(0, {ptvsd_path})
-import ptvsd
-from ptvsd.common import log, options
-del sys.path[0]
+    log_dir = log_dir.replace('\\', '/')
+    setup = {'host': host, 'port': port, 'client': client, 'log_dir': log_dir, 'pid': pid}
 
-options.log_dir = {log_dir}
-log.to_file()
-log.info("Bootstrapping injected debugger.")
+    # We need to be able to import attach_script_ptvsd_pid without importing ptvsd first.
+    if sys.platform == 'win32':
+        setup['pythonpath'] = attach_script_ptvsd_pid_dirname.replace('\\', '/')
 
-debug = ptvsd.attach if {client} else ptvsd.enable_attach
-debug(({host}, {port}), multiprocess={multiprocess})
-""".format(
-        **locals()
-    )
+        python_code = '''import sys;
+sys.path.insert(0, "%(pythonpath)s");
+import attach_script_ptvsd_pid;
+del sys.path[0];
+attach_script_ptvsd_pid.attach(port=%(port)s, host="%(host)s", client=%(client)s, log_dir="%(log_dir)s");
+'''.replace('\r\n', '').replace('\r', '').replace('\n', '')
+    else:
+        setup['pythonpath'] = attach_script_ptvsd_pid_dirname
+
+        # We have to pass it a bit differently for gdb
```

Comment:
It looks like the only difference here is escaping of quotes? If so, it would be better to have a single snippet, and just add backslashes if it's gdb. Or see if we can find out a way to avoid it - e.g. will single quotes work for both?

Reply:
Given that you'll probably need to resurrect `quote_str()`, it sounds like it would be easiest to just add that check there for the initial and the final `"` (and it will escape any inner ones).

Reply:
Yet another possibility - encode strings as lists of ints, each representing an UTF-8 byte:
```py
s = [ord(ch) for ch in s.encode("utf-8")]
```
That way, the injected snippet could have no quotes in it at all. On the receiving side (inside the attach script file), you can then do this:
```
s = bytearray(s).decode("utf-8")
```